Code refactoring

coffeemakr opened the pull request:
I thought a slightly bigger refactoring would do no harm. You can object to any changes and I can try to explain why I think it's a good change.

  * Add getters and setters to all field in the "model" classes InfoItem et al.
  * Add some javadocs
  * In Steam's use an MediaFormat object instead of an int which references the the id of the MediaFormat. 
  * Add an `ExtractorHelper` to remove duplicated code when extracting (related) streams using collectors.
  * Add a common interface for collectors (also to simplify the SearchCollector)
  * Add constructor parameters for *required* fields and make them final.

I've also ignored the Soundcloud Charts testGetName.

wb9688 commented:
I just took a _really_ quick look at it, and I agree with the changes. Btw could we improve NewPipeExtractor even more by using RxJava here too? I'm just asking…

theScrabi commented on src/main/java/org/schabi/newpipe/extractor/utils/ExtractorHelper.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.Collections;
+import java.util.List;
+
+public class ExtractorHelper {
```
What is this class for? Can't this be done by the collectors?

coffeemakr replied:
I've created this class to remove some duplicated code. But you're right this may be done in the collectors. I'll have a look at it.

theScrabi replied:
All right thank you :)

coffeemakr commented:
@wb9688 Yes I think the code and maybe the performance could be improved by parsing items asynchronously. For example parsing lists of videos could return a Flowable (or maybe Observable) so that streams can be added to the view as soon as it is parsed and also decouple computation and I/O.

theScrabi commented:
How comes no checks fail? Did you deactivate some tests or did you fix them?

coffeemakr commented:
It worked locally and remote I just ignored  the `testGetName` in  `SoundcloudChartsExtractorTest.java`

theScrabi commented:
Interesting for me some tests fail,  also if I load the html thr page loocks weird. But the app works correct.

Youtube is weird...

coffeemakr commented:
I've checked the project and it fails for me too.

coffeemakr commented:
* Rebased to be on top of the current master
 * I've added spotbugs library for `Nullable` and `NonNull` annotations this makes it easier to spot nullability issues.
 * Fixed issue generated by the assumption that `Map.get` throws an exception but instead returns null
 * Change some `assertTrue(!..)` to `assertFalse`
 * Ignore test to get audio streams for restricted YouTube video - Should this test return an empty list or not?
 * Ignore test to get the length for age restricted YouTube video - See #44

coffeemakr commented:
Sooo... I've found some more bugs:
  *  By introducing the #40 uploader url it was overlooked that there are anonymous extractors for playlists and channels (fixed in 9512064  d59618d0e729ddceec3245e02632edec1df70cf4)
  * Playlists can contain deleted items which should be skipped. I've implemented this logic in the `YoutubePlaylistExtractor` (commit d59618d0e729ddceec3245e02632edec1df70cf4)
 * Subtitles for restricted videos never worked but there are that's that fail. d76c8e1

coffeemakr commented:
The last commit doesn't seem to work.

theScrabi commented:
All right I fixed the few remaining errors :)

Well done :)

theScrabi commented:
No I didn't fix all issues, I only fixed the ones in the YoutubeStreamExtractorTest. The RestrictedYoutubeStreamExtractorTest still remains.
